MdeModulePkg/NonDiscoverablePciDeviceDxe: Avoid uncached bounce buffers

ardbiesheuvel opened the pull request:
The current handling of non-coherent DMA is very inefficient:

- outbound DMA uses bounce buffers unnecessarily, considering that a bus master read cannot break coherency, and can therefore access the host buffer directly without the risk of data corruption;

- bounce buffers are allocated from uncached memory, resulting in slow CopyMem() performance, and unnecessary overhead in allocating, remapping and freeing the pages, which also involves EFI memory map changes each time.

So refactor the logic, and tweak it so that:

- Bounce buffering is only performed when the host buffer is out of DMA range for the device, or when performing misaligned inbound DMA on a host buffer that is mapped cacheable;

- Cache maintenance is only performed when the buffer presented to the device is mapped cacheable.

- Bounce buffers are pool allocated, which is much more efficient and does not involve EFI memory map changes in the common case.

ardbiesheuvel committed:
MdeModulePkg/NonDiscoverablePciDeviceDxe: Avoid uncached bounce buffers

The current handling of non-coherent DMA is very inefficient:

- outbound DMA uses bounce buffers unnecessarily, considering that a bus
  master read cannot break coherency, and can therefore access the host
  buffer directly without the risk of data corruption;

- bounce buffers are allocated from uncached memory, resulting in slow
  CopyMem() performance, and unnecessary overhead in allocating,
  remapping and freeing the pages, which also involves EFI memory map
  changes each time.

So refactor the logic, and tweak it so that:

- Bounce buffering is only performed when the host buffer is out of DMA
  range for the device, or when performing misaligned inbound DMA on a
  host buffer that is mapped cacheable;

- Cache maintenance is only performed when the buffer presented to the
  device is mapped cacheable.

- Bounce buffers are pool allocated when possible, and combined with the
  MAP_INFO allocation: this is much more efficient and does not involve
  EFI memory map changes in the common case.

Signed-off-by: Ard Biesheuvel <ardb@kernel.org>

ardbiesheuvel commented:
@eeshanl

os-d commented on MdeModulePkg/Bus/Pci/NonDiscoverablePciDeviceDxe/NonDiscoverablePciDeviceIo.c:
```diff
@@ -1288,113 +1289,130 @@ NonCoherentPciIoMap (
+  // Check whether the host address refers to an uncached mapping. This is
+  // required for common buffer mappings.
+  //
+  Status = gDS->GetMemorySpaceDescriptor (
```
Github won't let me resolve the comment :)

ardbiesheuvel committed:
MdeModulePkg/NonDiscoverablePciDeviceDxe: Guard linked list with a lock

Avoid potential corruption of the uncached allocation list by using a
EFI_LOCK to guard it.

Signed-off-by: Ard Biesheuvel <ardb@kernel.org>

ardbiesheuvel committed:
MdeModulePkg/NonDiscoverablePciDeviceDxe: Avoid GCD for uncached check

Instead of relying on GCD services to decide whether or not a memory
region being mapped for DMA is mapped cacheable by the CPU, use the
allocation list which already exists to keep track of uncached buffers
that have been allocated on behalf of the device.

This is more reliable, and likely cheaper. Common buffer mappings are
only permitted for buffers that were allocated using the PCI I/O APIs in
the first place, and directional mappings can tolerate redundant cache
maintenance so in that case, misidentifying an uncached mapping as
cached is harmless.

Signed-off-by: Ard Biesheuvel <ardb@kernel.org>
